cli/internal/globby/globby.go

```diff
@@ -1,55 +1,141 @@
+func (s *set) Values() []string {
+	setValues := make([]string, len(s.storageMap))
+
+	i := 0
+	for key := range s.storageMap {
+		setValues[i] = key
+		i++
+	}
+
+	return setValues
+}
 
 func GlobFiles(basePath string, includePatterns []string, excludePatterns []string) []string {
-	return globFilesFs(_osFS, basePath, includePatterns, excludePatterns)
+	return globFilesFs(_aferoIOFS, basePath, includePatterns, excludePatterns)
 }
 
-func globFilesFs(fs afero.Fs, basePath string, includePatterns []string, excludePatterns []string) []string {
-	var include []string
-	var exclude []string
-	var result []string
+// getRelativePath ensures that the the requested file path is a child of `from`.
+func getRelativePath(from string, to string) (path string, err error) {
+	relativePath, err := filepath.Rel(from, to)
+
+	if err != nil {
+		return "", err
+	}
 
-	for _, p := range includePatterns {
-		include = append(include, filepath.Join(basePath, p))
+	if strings.HasPrefix(relativePath, "..") {
+		return "", errors.New("the path you are attempting to specify is outside of the root")
 	}
 
-	for _, p := range excludePatterns {
-		exclude = append(exclude, filepath.Join(basePath, p))
+	return relativePath, nil
+}
+
+// globFilesFs searches the specified file system to ensure to enumerate all files to include.
+func globFilesFs(fs afero.IOFS, basePath string, includePatterns []string, excludePatterns []string) []string {
+	var processedIncludes []string
+	var processedExcludes []string
+	var result = &set{
+		storageMap: map[string]bool{},
 	}
 
-	includePattern := "{" + strings.Join(include, ",") + "}"
-	excludePattern := "{" + strings.Join(exclude, ",") + "}"
+	for _, includePattern := range includePatterns {
+		includePath := filepath.Join(basePath, includePattern)
+		relativePath, err := getRelativePath(basePath, includePath)
+
+		if err == nil {
+			// Includes only operate on files.
+			processedIncludes = append(processedIncludes, relativePath)
+		}
+	}
 
-	_ = afero.Walk(fs, basePath, func(path string, info os.FileInfo, err error) error {
-		isDir := info.IsDir()
-		if val, _ := doublestar.PathMatch(excludePattern, path); val {
-			if isDir {
-				return filepath.SkipDir
-			}
+	for _, excludePattern := range excludePatterns {
+		excludePath := filepath.Join(basePath, excludePattern)
+		relativePath, err := getRelativePath(basePath, excludePath)
+
+		if err == nil {
+			// Excludes operate on entire folders.
+			processedExcludes = append(processedExcludes, filepath.Join(relativePath, "**"))
+		}
+	}
+
+	// We start from a naive includePattern
+	includePattern := ""
+	includeCount := len(processedIncludes)
+
+	// Do not use alternation if unnecessary.
+	if includeCount == 1 {
+		includePattern = filepath.Join(basePath, processedIncludes[0])
+	} else if includeCount > 1 {
+		// We start from a basePath prefix which allows doublestar to optimize the access.
+		includePattern = filepath.Join(basePath, "{"+strings.Join(processedIncludes, ",")+"}")
```

Comment:
This is a good catch of an assumption I made. Here is the full set of scenarios and their differences:

### Late Alternation (`/path/prefix/{foo/**,bar/**}`)
- For the case of a root path not on disk: exactly one stat
- For a successful case: `stat`s the alternation prefix

### Early Alternation (`{/path/prefix/foo/**,/path/prefix/bar/**}`)
- For the case of a root path not on disk: stats equal to alternation count.
- For a successful case: no alternation prefix to stat, so it does not stat

***

I believe I would prefer to limit downside at the cost of a single bonus `stat` but don't have strong opinions.

Reply:
I've updated the comment so that it matches the information uncovered in your research but I did not actually change the behavior. I'll defer to your call on which behavior you find to be better:

A. Optimize for the good case.
B. Limit the downside when garbage-in.

Reply:
Let's optimize for the good case. We're using `basePath` as the prefix, which is going to be either the repo root or a package root, and most likely we won't get this far if those don't exist, so I think the downside is both unlikely to occur and likely to be an error scenario if it does.

Reply:
Up in 34dbeb0557bf39b772e915934b7f051ba3a1315b.